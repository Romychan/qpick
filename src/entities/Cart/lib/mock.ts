// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
