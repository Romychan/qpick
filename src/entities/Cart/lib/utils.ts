// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
