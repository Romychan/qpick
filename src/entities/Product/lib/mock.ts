import { Product } from '../model/types';

/** List of mock products */
export const MOCK_PRODUCTS: Product[] = [
  {
    id: 1,
    title: 'Apple BYZ S852I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 2927,
    oldPrice: 3527,
    description:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    category: 'Наушники',
  },
  {
    id: 2,
    title: 'Apple AirPods Pro 3',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Наушники',
  },
  {
    id: 3,
    title: 'Apple EarPods',
    rate: 4.5,
    image: '/images/catalog/headphone-3.png',
    price: 2327,
    oldPrice: null,
    description:
      'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    category: 'Наушники',
  },
  {
    id: 4,
    title: 'Apple BYZ S123I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 1500,
    oldPrice: 1950,
    description:
      'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
    category: 'Наушники',
  },
  {
    id: 5,
    title: 'Apple AirPods Pro 2',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    category: 'Наушники',
  },
  {
    id: 6,
    title: 'Apple EarPods 2',
    rate: 3.8,
    image: '/images/catalog/headphone-3.png',
    price: 1200,
    oldPrice: 2000,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Наушники',
  },
  {
    id: 7,
    title: 'Apple AirPods',
    rate: 4.7,
    image: '/images/catalog/headphone-4.png',
    price: 9527,
    oldPrice: null,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Беспроводные наушники',
  },
  {
    id: 8,
    title: 'GERLAX GH-04',
    rate: 4.7,
    image: '/images/catalog/headphone-5.png',
    price: 6527,
    oldPrice: null,
    description:
      'Sed quis sem vitae justo feugiat tincidunt. Maecenas ut leo nisl. Praesent finibus ultricies ex et tristique.',
    category: 'Беспроводные наушники',
  },
  {
    id: 9,
    title: 'BOROFONE BO4',
    rate: 4.7,
    image: '/images/catalog/headphone-6.png',
    price: 7527,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Беспроводные наушники',
  },
];
