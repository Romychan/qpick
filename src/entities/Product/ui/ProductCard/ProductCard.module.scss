@import '~/app/assets/styles/_typography.scss';

.container {
  position: relative;
  display: flex;
  flex-direction: column;
  padding-bottom: 2rem;
}

.link {
  position: absolute;
  display: flex;
  align-items: center;
  justify-content: center;
  top: 1rem;
  right: 1rem;
  width: 28px;
  aspect-ratio: 1/1;
  background-color: #ffffff;
  border-radius: 8px;
  transition: 0.2s;

  & svg {
    stroke: var(--gray-600);
    transition: 0.2s;
  }

  &:hover {
    & svg {
      stroke: var(--yellow-500) !important;
    }
  }
}

.image {
  display: block;
  overflow: hidden;
  background: var(--background);
  max-width: 220px;
  align-self: center;

  & img {
    height: 100%;
    width: 100%;
    object-fit: contain;
    object-position: center;
  }
}

.information {
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 0.2rem;
  margin-top: 3.6rem;
}

.name {
  @include text-md;
  @include text-semibold;
  color: var(--gray-800);
  transition: 0.2s;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  &:hover {
    color: var(--yellow-500);
  }
}

.title {
  @include text-md;
  @include text-semibold;
  color: var(--text-gray-800);
  margin-bottom: 0.25rem;
  transition: 0.2s;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;

  &:hover {
    color: var(--yellow-500);
  }
}

.actions {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-top: 1.6rem;
  width: 100%;
  align-self: flex-end;
}
