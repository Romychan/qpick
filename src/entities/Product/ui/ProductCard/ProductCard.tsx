import { ReactNode } from 'react';
import { Link } from 'react-router-dom';

import { Rating } from '~/shared/ui/Rating';
import { Card } from '~/shared/ui/Card';
import { OVERVIEW_ROUTE } from '~/shared/config';
import { Icon } from '~/shared/ui/Icon';
import { Price } from '~/shared/ui/Price';

import { Product } from '../../model/types';

import styles from './ProductCard.module.scss';

interface IProductCardProps extends Product {
  /** The slot of the component for interactive elements */
  actions?: ReactNode;
}

/** This component is used to get brief information about the product in the form of a card */
export const ProductCard = ({
  id,
  title,
  image,
  rate,
  price,
  oldPrice,
  actions,
}: IProductCardProps) => {
  return (
    <Card className={styles.container} data-testid="product-card">
      <Link
        to={`${OVERVIEW_ROUTE}/${id}`}
        aria-label="Быстрый просмотр товара"
        title="Быстрый просмотр"
        className={styles.link}
      >
        <Icon name="eye" size={22} />
      </Link>

      <div className={styles.image}>
        <img src={image} width={220} height={238} alt={title} />
      </div>

      <div className={styles.information}>
        <Link to={`${OVERVIEW_ROUTE}/${id}`} className={styles.name}>
          {title}
        </Link>

        <Price
          price={price}
          oldPrice={oldPrice}
          currency="₽"
          direction="vertical"
        />
      </div>

      <div className={styles.actions}>
        <Rating value={rate} />
        {actions}
      </div>
    </Card>
  );
};
