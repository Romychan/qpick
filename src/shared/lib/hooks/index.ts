export { useEventListener } from './useEventListener';
export { useIsMounted } from './useIsMounted';
