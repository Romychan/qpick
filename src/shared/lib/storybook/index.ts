export { withReactHookFormDecorator } from './decorators/withReactHookFormDecorator';
export { withHeightDecorator } from './decorators/withHeightDecorator';
export { withGapDecorator } from './decorators/withGapDecorator';
export { withWidthDecorator } from './decorators/withWidthDecorator';
export { withRouterDecorator } from './decorators/withRouterDecorator';
