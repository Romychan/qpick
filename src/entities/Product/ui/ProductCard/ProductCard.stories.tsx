import type { Meta, StoryObj } from '@storybook/react';

import { Button } from '~/shared/ui/Button';
import {
  withRouterDecorator,
  withWidthDecorator,
} from '~/shared/lib/storybook';

import { ProductCard } from './ProductCard';

const meta = {
  title: 'Entities/Product/ProductCard',
  component: ProductCard,
  parameters: {
    layout: 'centered',
    width: '330px',
  },
  args: {
    id: 1,
    title: 'Product Name',
    image: './images/catalog/headphone-1.png',
    price: 1200,
    rate: 4.2,
  },
  decorators: [withRouterDecorator, withWidthDecorator],
  tags: ['autodocs'],
} satisfies Meta<typeof ProductCard>;

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to display the default */
export const Default: Story = {
  args: {
    actions: <Button>Actions Slot</Button>,
  },
};

/** This state is used to display the component with old and new price */
export const OldPrice: Story = {
  args: {
    oldPrice: 1600,
    actions: <Button>Actions Slot</Button>,
  },
};

/** Component without actions slot */
export const EmptySlots: Story = {
  args: {
    actions: null,
  },
};
