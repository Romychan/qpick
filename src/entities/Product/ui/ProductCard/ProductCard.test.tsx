import { screen } from '@testing-library/react';

import { renderWithRouter } from '~/shared/lib/tests';

import { ProductCard } from './ProductCard';

describe('ProductCard', () => {
  it('should render correctly', () => {
    renderWithRouter(
      <ProductCard
        id={1}
        title="Test"
        image="./test.png"
        price={1200}
        rate={4.1}
        actions={<button>Actions</button>}
      />,
    );

    const nameElement = screen.getByText(/Test/i);
    const ratingElement = screen.getByTestId('rating');
    const actionsElement = screen.getByRole('button');

    expect(nameElement).toBeInTheDocument();
    expect(ratingElement).toBeInTheDocument();
    expect(actionsElement).toBeInTheDocument();
  });

  it('should display the old price', () => {
    renderWithRouter(
      <ProductCard
        id={1}
        title="Test"
        image="./test.png"
        price={1200}
        rate={4.1}
        oldPrice={1500}
        actions={<button>Actions</button>}
      />,
    );

    const oldPriceElement = screen.getByText(/1500/i);

    expect(oldPriceElement).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const container = renderWithRouter(
      <ProductCard
        id={1}
        title="Test"
        image="./test.png"
        price={1200}
        rate={4.1}
        oldPrice={1500}
        actions={<button>Actions</button>}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
