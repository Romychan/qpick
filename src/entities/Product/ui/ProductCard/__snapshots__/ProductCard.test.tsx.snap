// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_1d0052 _container_01e065"
        data-testid="product-card"
      >
        <a
          aria-label="Быстрый просмотр товара"
          class="_link_01e065"
          href="/overview/1"
          title="Быстрый просмотр"
        >
          <svg
            class="icon icon-eye"
            data-testid="icon"
            height="22"
            width="22"
          >
            <use
              xlink:href="/images/icons.svg#eye"
            />
          </svg>
        </a>
        <div
          class="_image_01e065"
        >
          <img
            alt="Test"
            height="238"
            src="./test.png"
            width="220"
          />
        </div>
        <div
          class="_information_01e065"
        >
          <a
            class="_name_01e065"
            href="/overview/1"
          >
            Test
          </a>
          <div
            class="_container_5a3e99 _vertical_5a3e99 _md_5a3e99"
            data-testid="price"
          >
            <span
              class="_price_5a3e99"
            >
              1200
               
              ₽
            </span>
            <span
              class="_old_5a3e99"
            >
              1500
               
              ₽
            </span>
          </div>
        </div>
        <div
          class="_actions_01e065"
        >
          <div
            class="_container_861903"
            data-testid="rating"
          >
            <svg
              class="icon icon-star-filled _icon_861903 filled"
              data-testid="icon"
              height="24"
              width="24"
            >
              <use
                xlink:href="/images/icons.svg#star-filled"
              />
            </svg>
            <p
              class="_text_861903"
            >
              4.1
            </p>
          </div>
          <button>
            Actions
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_1d0052 _container_01e065"
      data-testid="product-card"
    >
      <a
        aria-label="Быстрый просмотр товара"
        class="_link_01e065"
        href="/overview/1"
        title="Быстрый просмотр"
      >
        <svg
          class="icon icon-eye"
          data-testid="icon"
          height="22"
          width="22"
        >
          <use
            xlink:href="/images/icons.svg#eye"
          />
        </svg>
      </a>
      <div
        class="_image_01e065"
      >
        <img
          alt="Test"
          height="238"
          src="./test.png"
          width="220"
        />
      </div>
      <div
        class="_information_01e065"
      >
        <a
          class="_name_01e065"
          href="/overview/1"
        >
          Test
        </a>
        <div
          class="_container_5a3e99 _vertical_5a3e99 _md_5a3e99"
          data-testid="price"
        >
          <span
            class="_price_5a3e99"
          >
            1200
             
            ₽
          </span>
          <span
            class="_old_5a3e99"
          >
            1500
             
            ₽
          </span>
        </div>
      </div>
      <div
        class="_actions_01e065"
      >
        <div
          class="_container_861903"
          data-testid="rating"
        >
          <svg
            class="icon icon-star-filled _icon_861903 filled"
            data-testid="icon"
            height="24"
            width="24"
          >
            <use
              xlink:href="/images/icons.svg#star-filled"
            />
          </svg>
          <p
            class="_text_861903"
          >
            4.1
          </p>
        </div>
        <button>
          Actions
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
